src/bridge/cockpitpcpmetrics.c

```diff
@@ -0,0 +1,1017 @@
+          a->scaleSpace == b->scaleSpace &&
+          a->dimCount == b->dimCount &&
+          a->dimTime == b->dimTime &&
+          a->dimSpace == b->dimSpace);
+}
+
+static gboolean units_convertible (pmUnits *a,
+                                   pmUnits *b)
+{
+  pmAtomValue dummy;
+  dummy.d = 0;
+  return pmConvScale (PM_TYPE_DOUBLE, &dummy, a, &dummy, b) >= 0;
+}
+
+static void
+cockpit_pcp_metrics_prepare (CockpitChannel *channel)
+{
+  CockpitPcpMetrics *self = COCKPIT_PCP_METRICS (channel);
+  const gchar *problem = "protocol-error";
+  JsonObject *options;
+  const gchar *source;
+  const gchar **instances = NULL;
+  const gchar **omit_instances = NULL;
+  JsonArray *metrics;
+  const char *name;
+  gint64 timestamp;
+  int type;
+  int i;
+
+  COCKPIT_CHANNEL_CLASS (cockpit_pcp_metrics_parent_class)->prepare (channel);
+
+  options = cockpit_channel_get_options (channel);
+
+  /* "source" option */
+  if (!cockpit_json_get_string (options, "source", NULL, &source))
+    {
+      g_warning ("invalid \"source\" option for metrics channel");
+      goto out;
+    }
+  else if (!source)
+    {
+      g_warning ("no \"source\" option specified for metrics channel");
+      goto out;
+    }
+  else if (g_str_has_prefix (source, "/"))
+    {
+      type = PM_CONTEXT_ARCHIVE;
+      name = source;
+    }
+  else if (g_str_equal (source, "direct"))
+    {
+      type = PM_CONTEXT_LOCAL;
+      name = NULL;
+    }
+  else if (g_str_equal (source, "pmcd"))
+    {
+      type = PM_CONTEXT_HOST;
+      name = "local:";
+    }
+  else
+    {
+      g_message ("unsupported \"source\" option specified for metrics: %s", source);
+      problem = "not-supported";
+      goto out;
+    }
+
+  self->name = source;
+  self->context = pmNewContext(type, name);
+  if (self->context < 0)
+    {
+      g_warning ("%s: couldn't create PCP context: %s", self->name, pmErrStr (self->context));
+      problem = "internal-error";
+      goto out;
+    }
+
+  /* "instances" option */
+  if (!cockpit_json_get_strv (options, "instances", NULL, (gchar ***)&instances))
+    {
+      g_warning ("%s: invalid \"instances\" option (not an array of strings)", self->name);
+      goto out;
+    }
+
+  /* "omit_instances" option */
+  if (!cockpit_json_get_strv (options, "omit_instances", NULL, (gchar ***)&omit_instances))
+    {
+      g_warning ("%s: invalid \"omit_instances\" option (not an array of strings)", self->name);
+      goto out;
+    }
+
+  /* "metrics" option */
+  self->numpmid = 0;
+  if (!cockpit_json_get_array (options, "metrics", NULL, &metrics))
+    {
+      g_warning ("%s: invalid \"metrics\" option was specified (not an array)", self->name);
+      goto out;
+    }
+  if (metrics)
+    self->numpmid = json_array_get_length (metrics);
+
+  self->pmidlist = g_new0 (pmID, self->numpmid);
+  self->metrics = g_new0 (MetricInfo, self->numpmid);
+  for (i = 0; i < self->numpmid; i++)
+    {
+      MetricInfo *info = &self->metrics[i];
```

Comment:
Suggest splitting this out into a separate function. This function is long enough already.